packages/treeview/src/useTreeview.ts

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { HTMLProps, useState } from 'react';
+import {
+  composeEventHandlers,
+  getControlledValue,
+  KEY_CODES
+} from '@zendeskgarden/container-utilities';
+import { useSelection } from '@zendeskgarden/container-selection';
+
+import {
+  handleArrowDown,
+  handleArrowLeft,
+  handleArrowRight,
+  handleArrowUp,
+  handleEnd,
+  handleHome
+} from './keyboardInteractions';
+import { IGetNodeProps, IGetTreeProps, IUseTreeviewProps, IUseTreeviewReturnValue } from './types';
+
+function requiredArguments(arg: any, argStr: string, methodName: string) {
+  if (typeof arg === 'undefined' || arg === null) {
+    throw new Error(
+      `Accessibility Error: You must provide an "${argStr}" option to "${methodName}()"`
+    );
+  }
+}
+
+const hasModifierKeyPressed = (e: KeyboardEvent): boolean =>
+  e.ctrlKey || e.metaKey || e.shiftKey || e.altKey;
+
+const SUPPORTED_KEYS = [
+  KEY_CODES.ENTER,
+  KEY_CODES.HOME,
+  KEY_CODES.END,
+  KEY_CODES.UP,
+  KEY_CODES.DOWN,
+  KEY_CODES.RIGHT,
+  KEY_CODES.LEFT
+];
+
+/**
+ * Custom hook to manage a TreeView component.
+ *
+ * https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView
+ *
+ */
+export function useTreeview<Item = any>({
+  openNodes,
+  horizontal,
+  rtl,
+  onChange = () => undefined,
+  ...options
+}: IUseTreeviewProps<Item> = {}): IUseTreeviewReturnValue<Item> {
+  const [ownFocusedItem, setFocusedItem] = useState<Item>();
+  const { selectedItem, focusedItem, getContainerProps, getItemProps } = useSelection<Item>({
+    direction: horizontal ? 'horizontal' : 'vertical',
+    defaultSelectedIndex: 0,
+    focusedItem: ownFocusedItem,
+    ...options
+  });
+  const [openNodesState, setOpenNodes] = useState<Item[]>([]);
+  const controlledOpenedState = getControlledValue<Item[] | undefined>(
+    openNodes,
+    Array.from(openNodesState)
+  )!;
+
+  const isControlled = openNodes !== null && openNodes !== undefined;
+  const isNodeExpanded = (item: Item | undefined): boolean => {
+    if (!item) {
+      return false;
+    }
+
+    return controlledOpenedState.includes(item);
+  };
+
+  const toggleParent = (item: Item) => {
+    const newValue = controlledOpenedState.includes(item)
+      ? controlledOpenedState.filter(i => i !== item)
+      : [...controlledOpenedState, item];
+
+    onChange(newValue);
+    if (!isControlled) {
+      setOpenNodes(newValue);
+    }
+  };
+
+  const getTreeProps = ({
+    label,
+    labelledBy,
+    role = 'tree',
+    onFocus,
+    ...props
+  }: IGetTreeProps = {}) => {
+    // TODO: Throw error if no label or labelledBy?
+    // TODO: Throw error role is not tree?
```

Comment:
As discussed on Slack we are going to throw errors when the prop values are different from what is specified in the W3 specs. 
This is a conscious deviation from Garden standard because the Treeview widget relies solely on attributes value to function.

Done in https://github.com/zendeskgarden/react-containers/pull/346/commits/4bc0c47b384cdbef15845688ac8d15accc5e90b8